cli/runner/orchestrator.go

```diff
@@ -0,0 +1,375 @@
+		logger:        logger,
+		openapiClient: openapiClient,
+		environments:  environments,
+	}
+}
+
+type orchestrator struct {
+	logger *zap.Logger
+
+	openapiClient *openapi.APIClient
+	environments  resourcemanager.Client
+}
+
+var (
+	yamlFormat = resourcemanager.Formats.Get(resourcemanager.FormatYAML)
+	jsonFormat = resourcemanager.Formats.Get(resourcemanager.FormatJSON)
+)
+
+const (
+	ExitCodeSuccess       = 0
+	ExitCodeGeneralError  = 1
+	ExitCodeTestNotPassed = 2
+)
+
+func (o orchestrator) Run(ctx context.Context, r Runner, opts RunOptions, outputFormat string) (exitCode int, _ error) {
+
+	o.logger.Debug(
+		"Running test from definition",
+		zap.String("definitionFile", opts.DefinitionFile),
+		zap.String("ID", opts.ID),
+		zap.String("envID", opts.EnvID),
+		zap.Bool("skipResultsWait", opts.SkipResultWait),
+		zap.String("junitOutputFile", opts.JUnitOuptutFile),
+	)
+
+	envID, err := o.resolveEnvID(ctx, opts.EnvID)
+	if err != nil {
+		return ExitCodeGeneralError, fmt.Errorf("cannot resolve environment id: %w", err)
+	}
+	o.logger.Debug("env resolved", zap.String("ID", envID))
+
+	var resource any
+	if opts.DefinitionFile != "" {
+		f, err := fileutil.Read(opts.DefinitionFile)
+		if err != nil {
+			return ExitCodeGeneralError, fmt.Errorf("cannot read definition file %s: %w", opts.DefinitionFile, err)
+		}
+		df := f
+		o.logger.Debug("Definition file read", zap.String("absolutePath", df.AbsPath()))
+
+		df, err = o.injectLocalEnvVars(ctx, df)
+		if err != nil {
+			return ExitCodeGeneralError, fmt.Errorf("cannot inject local env vars: %w", err)
+		}
+
+		resource, err = r.Apply(ctx, df)
+		if err != nil {
+			return ExitCodeGeneralError, fmt.Errorf("cannot apply definition file: %w", err)
+		}
+		o.logger.Debug("Definition file applied", zap.String("updated", string(df.Contents())))
+	} else {
+		o.logger.Debug("Definition file not provided, fetching resource by ID", zap.String("ID", opts.ID))
+		resource, err = r.GetByID(ctx, opts.ID)
+		if err != nil {
+			return ExitCodeGeneralError, fmt.Errorf("cannot get resource by ID: %w", err)
+		}
+		o.logger.Debug("Resource fetched by ID", zap.String("ID", opts.ID), zap.Any("resource", resource))
+	}
+
+	var result RunResult
+	var ev envVars
+
+	// iterate until we have all env vars,
+	// or the server returns an actual error
+	for {
+		runInfo := openapi.RunInformation{
+			EnvironmentId: &envID,
+			Variables:     ev.toOpenapi(),
+			Metadata:      getMetadata(),
+		}
+
+		result, err = r.StartRun(ctx, resource, runInfo)
+		if err == nil {
+			break
+		}
+		if !errors.Is(err, missingEnvVarsError{}) {
+			// actual error, return
+			return ExitCodeGeneralError, fmt.Errorf("cannot run test: %w", err)
+		}
+
+		// missing vars error
+		ev = askForMissingVars([]envVar(err.(missingEnvVarsError)))
+		o.logger.Debug("filled variables", zap.Any("variables", ev))
+	}
+
+	if opts.SkipResultWait {
+		fmt.Println(r.FormatResult(result, outputFormat))
+		return ExitCodeSuccess, nil
+	}
```

Comment:
(Something to discuss with the team) 

I think that we should tell the user that the test is running and there are no guarantees that it will pass, even knowing that now the user should explicitly pass an option to avoid waiting for the test result. Does it make sense?

Today, by doing the standard formatting our runner will just print a "header" of the Test output, which can make our users think that this test passed.

Reply:
tracking on [this issue](https://github.com/kubeshop/tracetest/issues/2930)